packages/grid/x-data-grid-pro/src/hooks/features/columnResize/useGridColumnResize.tsx

```diff
@@ -426,12 +518,168 @@ export const useGridColumnResize = (
+      if (props.disableAutosize) {
+        return;
+      }
+
+      // Only handle left clicks
+      if (event.button !== 0) {
+        return;
+      }
+
+      const column = apiRef.current.state.columns.lookup[params.field];
+      const options = { ...props.autosizeOptions };
+      options.columns ??= [column];
+
+      apiRef.current.autosizeColumns(options);
+    });
+
+  /**
+   * API METHODS
+   */
+
+  const columnVirtualizationDisabled = useColumnVirtualizationDisabled(apiRef);
+  const isAutosizingRef = React.useRef(false);
+  const autosizeColumns = React.useCallback<GridColumnResizeApi['autosizeColumns']>(
+    async (userOptions) => {
+      const root = apiRef.current.rootElementRef?.current;
+      if (!root) {
+        return;
+      }
+      if (isAutosizingRef.current) {
+        return;
+      }
+      isAutosizingRef.current = true;
+
+      const state = gridColumnsStateSelector(apiRef.current.state);
+      const options = {
+        ...DEFAULT_GRID_AUTOSIZE_OPTIONS,
+        ...userOptions,
+        columns: userOptions?.columns ?? state.orderedFields.map((field) => state.lookup[field]),
+      };
+      options.columns = options.columns.filter(
+        (c) => state.columnVisibilityModel[c.field] !== false,
+      );
+
+      try {
+        apiRef.current.unstable_setColumnVirtualization(false);
+        await columnVirtualizationDisabled();
+
+        const widthByField = extractColumnWidths(apiRef, options);
+
+        const newColumns = options.columns.map((column) => ({
+          ...column,
+          width: widthByField[column.field],
+          computedWidth: widthByField[column.field],
+        }));
+
+        if (options.expand) {
+          const visibleColumns = state.orderedFields
+            .map((field) => state.lookup[field])
+            .filter((c) => state.columnVisibilityModel[c.field] !== false);
+
+          const totalWidth = visibleColumns.reduce(
+            (total, column) =>
+              total + (widthByField[column.field] ?? column.computedWidth ?? column.width),
+            0,
+          );
+          const availableWidth = apiRef.current.getRootDimensions()?.viewportInnerSize.width ?? 0;
+          const remainingWidth = availableWidth - totalWidth;
+
+          if (remainingWidth > 0) {
+            const widthPerColumn = remainingWidth / (newColumns.length || 1);
+            newColumns.forEach((column) => {
+              column.width += widthPerColumn;
+              column.computedWidth += widthPerColumn;
+            });
+          }
+        }
+
+        apiRef.current.updateColumns(newColumns);
+      } finally {
+        apiRef.current.unstable_setColumnVirtualization(true);
+        isAutosizingRef.current = false;
+      }
+    },
+    [apiRef, columnVirtualizationDisabled],
+  );
+
+  /**
+   * EFFECTS
+   */
+
+  React.useEffect(() => stopListening, [stopListening]);
+
+  useOnMount(() => {
+    if (props.autosizeOnMount) {
+      Promise.resolve().then(() => {
```

Comment:
If I understand it correctly, `Promise` here pushes the start of autosizing process to end of call stack to make sure the grid is mounted completely beforehand, right?

Reply:
Yes, we have other parts of the grid that use effects to propagate their changes, e.g.

```tsx
useEffect(() => {
  setLocalColumnsState(props.columns)
}, [props.columns])
```

Reply:
Yes, we are using effects but we are not using a `Promise.resolve` elsewhere, or do we 🤔

Reply:
I'm lazy. I'm not doing it unless you think it should really be done that way. Lmk :)

Reply:
I think it should be fine this way, I was just curious about the need for `Promise.resolve()` here.